# Bag-of-words transformations of our TMDB and IMDB movie plots

This notebook applies a bag-of-words with TFIDF transformation to our TMDB and IMDB movie plots. Further design considerations and analysis are below.

In [1]:
#import libraries
import pandas as pd
from ast import literal_eval
from nltk.tokenize import RegexpTokenizer
from stop_words import get_stop_words
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.feature_extraction.text import TfidfVectorizer
import numpy as np


In [2]:
#list of columns in our 'movies.csv' that need to be read in using literal_eval
#to ensure they are represented as vectors rather than strings
converterList = ['imdb_genres', 'tmdb_genres', 'binary_tmdb', 'binary_imdb',
               'imdb_w2v_plot', 'tmdb_w2v_plot']

converterDict = {column: literal_eval for column in converterList}

movies = pd.read_csv('data/movies.csv', encoding='utf-8',
                     converters=converterDict)

movies.head(5)

tmdb_id    imdb_id      tmdb_genres         imdb_genres  \
0      278  tt0111161         [18, 80]            [80, 18]   
1      238  tt0068646         [18, 80]            [80, 18]   
2      424  tt0108052  [18, 36, 10752]            [18, 36]   
3      240  tt0071562         [18, 80]            [80, 18]   
4   452522  tt0278784       [18, 9648]  [80, 18, 9648, 53]   

                                         binary_tmdb  \
0  [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, ...   
1  [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, ...   
2  [0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, ...   
3  [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, ...   
4  [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, ...   

                                         binary_imdb  \
0  [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, ...   
1  [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, ...   
2  [0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, ...   
3  [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, ...   
4  [0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, ...   

                                           tmdb_plot  \
0  Framed in the 1940s for the double murder of h...   
1  Spanning the years 1945 to 1955  a chronicle o...   
2  The true story of how businessman Oskar Schind...   
3  In the continuing saga of the Corleone crime f...   
4  Standalone version of the series pilot with an...   

                                           imdb_plot  popularity release_date  \
0  Chronicles the experiences of a formerly succe...   28.527767   1994-09-23   
1  When the aging head of a famous crime family d...   36.965452   1972-03-14   
2  Oskar Schindler is a vainglorious and greedy G...   19.945455   1993-11-29   
3  The continuing saga of the Corleone crime fami...   30.191804   1974-12-20   
4  When beautiful  young Laura Palmer is found br...    5.969249   1989-12-31   

                         ...                         vote_average  vote_count  \
0                        ...                                  8.5        9773   
1                        ...                                  8.5        7394   
2                        ...                                  8.4        5518   
3                        ...                                  8.4        4249   
4                        ...                                  8.4         123   

                                     tmdb_clean_plot  \
0  ['framed', '1940s', 'double', 'murder', 'wife'...   
1  ['spanning', 'years', '1945', '1955', 'chronic...   
2  ['true', 'story', 'businessman', 'oskar', 'sch...   
3  ['continuing', 'saga', 'corleone', 'crime', 'f...   
4  ['standalone', 'version', 'series', 'pilot', '...   

                                     imdb_clean_plot  \
0  ['chronicles', 'experiences', 'formerly', 'suc...   
1  ['aging', 'head', 'famous', 'crime', 'family',...   
2  ['oskar', 'schindler', 'vainglorious', 'greedy...   
3  ['continuing', 'saga', 'corleone', 'crime', 'f...   
4  ['beautiful', 'young', 'laura', 'palmer', 'fou...   

                                       tmdb_w2v_plot  \
0  [0.014165705069899559, 0.035729147493839264, 0...   
1  [-0.016820836812257767, 0.05966977775096893, -...   
2  [0.0758906751871109, 0.02254812978208065, 0.06...   
3  [-0.05790800228714943, 0.07111673057079315, -0...   
4  [-0.05888228118419647, -0.05345569923520088, -...   

                                       imdb_w2v_plot  \
0  [0.004663567990064621, 0.09018586575984955, -0...   
1  [-0.013326308690011501, 0.08134819567203522, 0...   
2  [0.05338115245103836, 0.10281133651733398, 0.0...   
3  [-0.05151921883225441, 0.07896284759044647, -0...   
4  [0.0026558770332485437, 0.10140948742628098, 0...   

                                       tmdb_bow_plot  \
0    (0, 700)\t0.1914190824267342\n  (0, 1141)\t0...   
1    (0, 610)\t0.12626163649598618\n  (0, 1165)\t...   
2    (0, 1079)\t0.2907037443234484\n  (0, 990)\t0...   
3    (0, 720)\t0.19462813339213522\n  (0, 243)\t0...   
4    (0, 1110)\t0

---
### Make new feature of combined TMDB and IMDB plots
We will make a new column with the combined plots of tmdb and imdb. Perhaps this will offer a richer representation of each plot and increase our precision and recall in the forthcoming modeling.

In [3]:
movies['combined_plots'] = movies['tmdb_plot'] + ' ' + movies['imdb_plot']

---
### Apply Count Vectorize and TFIDF to Plots

Sklearn's TfidfVectorizer will apply both a count vectorizer and Term Frequency Inverse Document Frequency (TFIDF) transformation to our plots. The count vectorizer will convert our collection of plots into a matrix of word counts. We set max_df and min_df to ignore words specific to our corpus of movie plots that occur both a lot and few times, and we also ignore english stop words if the previous arguments missed some. The TFIDF transformation will essentially apply a rarity score to each word returning a normalized TFIDF matrix for each feature - TMDB, IMDB, and combined plots.

In [4]:
tfidf_vectorizer = TfidfVectorizer(max_df=0.95, min_df=0.005, 
                          stop_words = get_stop_words('en'))

tmdb_bow=tfidf_vectorizer.fit_transform(movies['tmdb_plot'])
imdb_bow=tfidf_vectorizer.fit_transform(movies['imdb_plot'])
combined_bow = tfidf_vectorizer.fit_transform(movies['combined_plots'])

#check shape of resulting matrices
tmdb_bow.shape, imdb_bow.shape, combined_bow.shape

((1000, 1171), (1000, 2442), (1000, 3025))

As seen above, now each TMDB and IMDB movies are represented by an 1171 and 2439 length array, and each combined plot is a 3025 length array, respectively. Each number in the array is the associated TFIDF score for the count vectorized representation of each plot. Most of the array values will be zero since each movie has on average 50 and 100 words for TMDB and IMDB, respectively.

---

### Add Bag-of-Words Matrices to Movies DF and save

We cast the resulting arrays into lists that pandas can save in a nice fashion for when we import in the modeling and analysis notebook. We also save them as numpy arrays for easier importing.

In [5]:
#each plot is a row vector of the TFIDF sparse matrices
movies['tmdb_bow_plot'] = [plot for plot in tmdb_bow]
movies['imdb_bow_plot'] = [plot for plot in imdb_bow]
movies['combined_bow_plots'] = [plot for plot in combined_bow]

#resave movies csv
movies.to_csv('data/movies.csv', encoding="utf-8", index=False)

#save TFIDF matrices an numpy arrays
np.save('data/tmdb_bow.npy',tmdb_bow.toarray())
np.save('data/imdb_bow.npy',imdb_bow.toarray())
np.save('data/combined_bow.npy', combined_bow.toarray())

### We have now transformed and saved each plot as a bag-of-words TFIDF representation.